File: NaviTest/NaviTest/PinkieNavigation/PinkieNavigationView.m
//
//  PinkieNavigationView.m
//  NaviTest
//
//  Created by jiazhiyi on 2019/4/9.
//  Copyright © 2019 jiazhiyi. All rights reserved.
//

#import "PinkieNavigationView.h"

@interface PinkieNavigationView ()

@property (nonatomic, strong) NSMutableArray  *itemArray;

@end

@implementation PinkieNavigationView

- (void)addNavigationItem:(PinkieNavigationItem *)item
{
    if ([self.itemArray containsObject:item]) {
        return;
    }
    
    PinkieBarButtonView *leftItemView = [item.leftBarButtonItem getBarButtonView];
    CGFloat leftWidth = [leftItemView viewWidth];
    leftItemView.frame = CGRectMake(8.0, self.frame.size.height - 44.0, leftWidth, 44.0);
    [self addSubview:leftItemView];
    
    PinkieBarButtonView *rightItemView = [item.rightBarButtonItem getBarButtonView];
    CGFloat rightWidth = [rightItemView viewWidth];
    rightItemView.frame = CGRectMake(self.frame.size.width - rightWidth - 8.0, self.frame.size.height - 44, rightWidth, 44.0);
    [self addSubview:rightItemView];
    
    CGFloat width = self.frame.size.width - MAX(leftWidth, rightWidth) * 2.0 - 16.0;
    CGRect rect = CGRectMake(0.0, 0.0, width, 44.0);
    UILabel *label = [[UILabel alloc] initWithFrame:rect];
    label.center = CGPointMake(self.frame.size.width / 2.0, self.frame.size.height - 22.0);
    label.textAlignment = NSTextAlignmentCenter;
    label.text = item.title;
    [self addSubview:label];
    item.titleLabel = label;
    [self.itemArray addObject:item];
}

- (void)removeNavigationItem:(PinkieNavigationItem *)item
{
    if ([self.itemArray containsObject:item]) {
        [item.titleLabel removeFromSuperview];
        [[item.leftBarButtonItem getBarButtonView] removeFromSuperview];
        [[item.rightBarButtonItem getBarButtonView] removeFromSuperview];
        [self.itemArray removeObject:item];
    }
}

- (void)removeAllItems
{
    for (PinkieNavigationItem *item in self.itemArray) {
        [item.titleLabel removeFromSuperview];
        [[item.leftBarButtonItem getBarButtonView] removeFromSuperview];
        [[item.rightBarButtonItem getBarButtonView] removeFromSuperview];
    }
    [self.itemArray removeAllObjects];
}

- (NSMutableArray *)itemArray
{
    if (_itemArray == nil) {
        _itemArray = [NSMutableArray arrayWithCapacity:2];
    }
    return _itemArray;
}

@end

File: CustomNavigationBar/CustomNavigationBar/PinkieNavigation/PinkieNavigationView.swift
//
//  PinkieNavigationView.swift
//  CustomNavigationBar
//
//  Created by jiazhiyi on 2019/4/4.
//  Copyright © 2019 jiazhiyi. All rights reserved.
//

import UIKit

class PinkieNavigationView: UIView {
    
    var itemArray = NSMutableArray()

    func addNavigationItem(item: PinkieNavigationItem?) {
        if let item = item {
            if !self.itemArray.contains(item) {
                
                var leftWidth = 8.0
                if let leftItemView = item.leftBarButtonItem?.getBarButtonView() {
                    leftWidth = leftItemView.viewWidth()
                    leftItemView.frame = CGRect(x: 8.0, y: Double(self.frame.height - 44.0), width: leftWidth, height: 44.0)
                    self.addSubview(leftItemView)
                }
                
                var rightWidth = 8.0
                if let rightItemView = item.rightBarButtonItem?.getBarButtonView() {
                    rightWidth = rightItemView.viewWidth()
                    rightItemView.frame = CGRect(x: Double(self.frame.width) - rightWidth - 8.0, y: Double(self.frame.height - 44), width: rightWidth, height: 44.0)
                    self.addSubview(rightItemView)
                }
                
                let width = Double(self.frame.width) - max(leftWidth, rightWidth) * 2.0 - 16.0
                let label = UILabel(frame: CGRect(x: 0.0, y: 0.0, width: width, height: 44.0))
                label.center = CGPoint(x: self.frame.width / 2.0, y: self.frame.height - 22.0)
                label.textAlignment = .center
                label.text = item.title
                self.addSubview(label)
                item.titleLabel = label
                self.itemArray.add(item)
            }
        }
    }

    func removeNavigationItem(item: PinkieNavigationItem?) {
        if let item = item {
            if self.itemArray.contains(item) {
                item.titleLabel?.removeFromSuperview()
                item.leftBarButtonItem?.getBarButtonView().removeFromSuperview()
                item.rightBarButtonItem?.getBarButtonView().removeFromSuperview()
                self.itemArray.remove(item)
            }
        }
    }
    
    func removeAllItems() {
        self.itemArray.forEach { (item) in
            if let item = item as? PinkieNavigationItem {
                item.titleLabel?.removeFromSuperview()
                item.leftBarButtonItem?.getBarButtonView().removeFromSuperview()
                item.rightBarButtonItem?.getBarButtonView().removeFromSuperview()
            }
        }
        self.itemArray.removeAllObjects()
    }
}
